pkg/cmd/cli/describe/chaindescriber.go

```diff
@@ -0,0 +1,267 @@
+					if err != nil {
+						return err
+					}
+					avoidDuplicates[stream] = struct{}{}
+				default:
+					// ImageStreamImage and DockerImage are never updated and so never
+					// trigger builds.
+					continue
+				}
+
+			}
+		}
+	}
+
+	for stream := range avoidDuplicates {
+		l.items = append(l.items, *stream)
+	}
+
+	return nil
+}
+
+// dotDump traverses the given graph starting from the provided stream node
+// and returns the stream's dependency tree in DOT syntax
+func dotDump(g osgraph.Graph, stream graph.Node, tag string) (string, error) {
+	dotGraph := dot.NewGraph()
+	attrs := make(map[string]string)
+
+	// Set graph name based on the root node
+	name := g.GraphDescriber.Name(stream)
+	dotGraph.SetName(validDOT(name))
+	// Directed graph since we illustrate dependencies
+	dotGraph.SetDir(true)
+	// Explicitly allow multiple pairs of edges between
+	// the same pair of nodes
+	dotGraph.SetStrict(false)
+	// Add root node
+	kind := g.GraphDescriber.Kind(stream)
+	attrs["label"] = fmt.Sprintf("<%s<BR /><FONT POINT-SIZE=\"10\">%s</FONT>>", name, kind)
+	dotGraph.AddNode(dotGraph.Name, g.GraphDescriber.Name(stream), attrs)
+
+	// Traverse the graph. Build the DOT graph along the way.
+	traverse(g, dotGraph, stream, tag, attrs)
+
+	// Parse DOT output for validation
+	out := dotGraph.String()
+	if _, err := dot.Parse([]byte(out)); err != nil {
+		return "", fmt.Errorf("cannot parse DOT output: %v", err)
+	}
+	return out, nil
+}
+
+// validDOT replaces hyphens with undescores so to
+// keep the DOT parser happy
+func validDOT(name string) string {
+	// TODO: The only special character the DOT parser
+	// accepts is the underscore (_)
+	return strings.Replace(name, "-", "_", -1)
+}
+
+// traverse will first look in the predecessors of an image stream.
+// It is expected to find imageStreamTags. Depending on finding the
+// right tag, it will continue via successor BuildInputImage edges
+// until it finds the correct buildConfig which contains the trigger
+// we are interested in. Via that trigger it will check for successor
+// BuildOutput edges leading in an imageStreamTag which in turn will
+// point us to an imageStream for our dep graph.
+func traverse(g osgraph.Graph, dotGraph *dot.Graph, stream graph.Node, tag string, attrs map[string]string) error {
+	// First look for imageStreamTags in the predecessors of the provided stream
+	var tagNode graph.Node
+	for _, streamTag := range g.Predecessors(stream) {
+		if g.GraphDescriber.Name(streamTag) == tag {
+			tagNode = streamTag
+
+			// Create tag node in the DOT graph
+			name := g.GraphDescriber.Name(tagNode)
+			kind := g.GraphDescriber.Kind(tagNode)
+			attrs["label"] = fmt.Sprintf("<%s<BR /><FONT POINT-SIZE=\"10\">%s</FONT>>", name, kind)
+			dotGraph.AddNode(dotGraph.Name, name, attrs)
+			// Create DOT edge
+			dotGraph.AddEdge(g.GraphDescriber.Name(stream), name, true, nil)
+			break
+		}
+	}
+	if tagNode == nil {
+		// TODO: I somehow have to have synthetic nodes here (?)
```

Comment:
I don't think that synthetics matter to you, since all bc outputs will be realized.